File: android/src/main/kotlin/fr/dooz/nordic_nrf_mesh/DoozUnprovisionedMeshNode.kt
package fr.dooz.nordic_nrf_mesh

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import no.nordicsemi.android.mesh.provisionerstates.UnprovisionedMeshNode

class DoozUnprovisionedMeshNode(binaryMessenger: BinaryMessenger, var meshNode: UnprovisionedMeshNode) :MethodChannel.MethodCallHandler {
    init {
        MethodChannel(binaryMessenger, "$namespace/unprovisioned_mesh_node/${meshNode.deviceUuid}/methods").setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getNumberOfElements" -> {
                result.success(meshNode.numberOfElements)
            }
            "setUnicastAddress" -> {
                meshNode.unicastAddress = call.argument<Int>("unicastAddress")!!
                result.success(null)
            }
        }
    }
}

File: ios/Classes/DoozUnprovisionedMeshNode/DoozUnprovisionedMeshNode.swift
//
//  DoozUnprovisionedMeshNode.swift
//  nordic_nrf_mesh
//
//  Created by Alexis Barat on 06/07/2020.
//

import Foundation
import nRFMeshProvision

class DoozUnprovisionedMeshNode: NSObject{
    
    //MARK: Public properties
    var unprovisionedMeshNode: UnprovisionedDevice?
    
    //MARK: Private properties
    private var eventSink: FlutterEventSink?

    init(messenger: FlutterBinaryMessenger, unprovisionedMeshNode: UnprovisionedDevice) {
        super.init()
        self.unprovisionedMeshNode = unprovisionedMeshNode
        _initChannels(messenger: messenger, unprovisionedMeshNode: unprovisionedMeshNode)
    }
    
    
}

private extension DoozUnprovisionedMeshNode {
    
    func _initChannels(messenger: FlutterBinaryMessenger, unprovisionedMeshNode: UnprovisionedDevice){

        FlutterMethodChannel(
            name: FlutterChannels.DoozUnprovisionedMeshNode.getMethodChannelName(deviceUUID: unprovisionedMeshNode.uuid.uuidString),
            binaryMessenger: messenger
        )
            .setMethodCallHandler { (call, result) in
                self._handleMethodCall(call, result: result)
        }
        
    }
    
    
    func _handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        print("🥂 [\(self.classForCoder)] Received flutter call : \(call.method)")
        
        guard let _method = DoozUnprovisionedMeshNodeChannel(rawValue: call.method) else{
            print("❌ Plugin method - \(call.method) - isn't implemented")
            return
        }
        
        switch _method {
        
        case .getNumberOfElements:
            
            break
            
        case .setUnicastAddress:
            #warning("to implement or remove if useless in ios")
            break
        }

    }
    
}

//private extension DoozUnprovisionedMeshNode{
//    // Events native implemenations
//    
//    func _getMeshNetworkName() -> String?{
//        return meshNetwork?.meshName
//    }
//    
//    func _getId() -> String?{
//        return meshNetwork?.id
//    }
//
//}
